### ISSUE TITLE
When native is True and reload is True, port is ignored and app_shutdown never fires

### ISSUE THREAD

**Issue by PawelRoman** [zauberzeug/nicegui#2107]
### Description

Platform:
Linux Mint 20.03 (ubuntu)
NiceGUI 1.4.3

Steps to replicate:
Save the code below as main.py and run `python main.py`

```py
from nicegui import app, ui

def app_startup():
    print('app startup')

def app_shutdown():
    print('app shutdown')

app.on_startup(app_startup)
app.on_shutdown(app_shutdown)

ui.run(
    title="Some app",
    port=8502,
    native=True,
    reload=True,
    window_size=(1920, 1200),
    storage_secret='my secret'
)
```

Expected: Window app opens up, server starts on port 8502, app startup printed out.
Actual:  Window app opens up, app startup printed out, but server starts on port 8000 (note: I double checked that 8502 is NOT taken by some other process)

Then close the window by clicking "X"

Expected: app shutdown printed out
Actual: nothing is printed out, the app_shutdown function is never called

**Comment by retsyo** [comment]
Currently, port is not used in native mode as https://github.com/zauberzeug/nicegui/discussions/1457 stated. I have made a [pull request](https://github.com/zauberzeug/nicegui/pull/2110) to change this.

However, I don't know why app_shutdown never fires.

**Comment by rodja** [comment]
Thanks @retsyo. On Mac M1 the shutdown works as expected. Are you running on Windows @PawelRoman? Does the shutdown work if you disable reload?

**Comment by retsyo** [comment]
On my windows 10 64 bits, both buttons close the window and python 
```python
ui.button('destroy window', on_click=lambda: app.native.main_window.destroy())
ui.button('nicegui shutdown', on_click=lambda: app.shutdown())

ui.run(reload=False, native=True)
```

and [PawelRoman](https://github.com/PawelRoman)'s code runs `app_shutdown` too if `reload=False`
```
def app_shutdown():
    print('app shutdown')

app.on_shutdown(app_shutdown)
ui.run(reload=False, native=True)
```

so does `reload=True` conflict with `app.on_shutdown`?

**Comment by rodja** [comment]
@retsyo are both buttons not working when using `reload=True, native=True`?

**Comment by rodja** [comment]
I fear this must be debugged on a Windows system. On Mac the demo you provided prints the shutdown message. Any help would be appreciated.

**Comment by PawelRoman** [comment]
As I wrote in the original post, I work on Ubuntu 20 (Linux Mint). 

The `on_shutdown` issue gets fixed when I set reload to False, with reload=True, the `on_shutdown` does not fire at all. Would also be great if the port issue fix be merged asap, as it's very annoying in certain scenarios not being able to select specific port.

**Comment by frankhuurman** [comment]
@PawelRoman thanks for the port fix :) saw that it got merged

Tried your examples on a Win 10 machine & python 3.11, came to the same conclusion that the app startup does get printed and the app shutdown doesn't fire with reload=True unless I Ctrl+C the app, when I set reload=False it fires without having to use Ctrl+C.

So that means on both Linux Mint & Win 10 the shutdown event doesn't seem to trigger with reload set to True.

Perhaps the reload/directory&file watcher functionality makes uvicorn spawn a watcher thread and won't let go of it unless Ctrl+C'd?

**Comment by falkoschindler** [comment]
Uvicorn just fixed https://github.com/encode/uvicorn/issues/2289 and released version 0.30.0.
Can anyone check if this by any chance also fixes our issue?

**Comment by falkoschindler** [comment]
@07rohanraj just solved a similar issue [like this](https://github.com/zauberzeug/nicegui/discussions/4181#discussioncomment-11819633):
```py
def shutdown_app():
    print("Shutting down the application...")
    logging.info("Shutting down the application...")
    sys.exit(0)
 
app.on_shutdown(shutdown_app)

ui.run(native=True, window_size=(1400, 800), reload=False, port=native.find_open_port())
```

Can anyone confirm this is working for others as well? Maybe this is a reliable workaround.

**Comment by falkoschindler** [comment]
Is there anyone still around to check

- if the problem still persists,
- if the uvicorn release mentioned in https://github.com/zauberzeug/nicegui/issues/2107#issuecomment-2134640373 helps, or
- if the workaround mentioned in https://github.com/zauberzeug/nicegui/issues/2107#issuecomment-2587561324 helps?

Otherwise we will close this issue.

**Comment by krashdifferent** [comment]
@falkoschindler I just ran into this issue or a very similar issue of closing the native window not closing the python. This is running in windows. I tried several different environments as I swear this use to work as expected for me.

Python 3.13.2
nicegui 2.13.0
uvicorn 0.34.0

Python 3.11.7
nicegui 2.12.1
uvicorn 0.25.0 and 0.30.0 and 0.34.0

The server does not appear to shutdown at all, as starting a second copy connects to the original server. This holds true for applications complied to exe in windowed mode with reload=False as well, which has been a problem without a user-friendly way to kill the background process.

**Comment by falkoschindler** [comment]
> However, with reload=True, app_shutdown is never called, independent of unicorn version (0.34.0, 0.30.0, 0.29.0, or 0.25.0) and the server process stays active consentient with the original issue reported above.

Ok, so the original problem persists: The shutdown handler is not called when running with `native=True` and `reload=True`.

**Comment by evnchn** [comment]
`app.on_shutdown`, eh? 
* #4641 

Try moving to 2.16.0 and see if it fixes things.

**Comment by python-and-novella** [comment]
NiceGUI 3.0.0, Windows 11 ltsc, Python 3.13.7,
We can see `'app shutdown'` when click button `'nicegui shutdown'` or use `ctrl+c`:

```python3
from nicegui import app, ui

def app_startup():
    print('app startup')

def app_shutdown():
    print('app shutdown')

app.on_startup(app_startup)
app.on_shutdown(app_shutdown)

@ui.page('/')
def index():
    ui.button('destroy window', on_click=lambda: app.native.main_window.destroy())
    ui.button('nicegui shutdown', on_click=lambda: app.shutdown())
    
ui.run(
    title="Some app",
    port=8502,
    native=True,
    reload=True,
    window_size=(1920, 1200),
    storage_secret='my secret'
)
```

**Comment by falkoschindler** [comment]
@python-and-novella Sounds great! Does it also work when clicking on the "X" button as @PawelRoman described in the original post? If yes, I'd suggest to close this issue. It's probably been solved upstream.

**Comment by python-and-novella** [comment]
@falkoschindler actually, We can ONLY see `'app shutdown'` when click button `'nicegui shutdown'` or use `ctrl+c`.
If we click `'X'` button or button `'destroy window'`, app is still running.

**Comment by python-and-novella** [comment]
I get a temporay solution by adding `app.shutdown` into `app.on_disconnect`:
```python3
from nicegui import app, ui

def app_startup():
    print('app startup')

def app_shutdown():
    print('app shutdown')

app.on_startup(app_startup)
app.on_shutdown(app_shutdown)
app.on_disconnect(app.shutdown)

@ui.page('/')
def index():
    ui.button('destroy window', on_click=lambda: app.native.main_window.destroy())
    ui.button('nicegui shutdown', on_click=lambda: app.shutdown())
    
ui.run(
    title="Some app",
    port=8502,
    native=True,
    reload=True,
    window_size=(1920, 1200),
    storage_secret='my secret'
)
```

**Comment by Ansari-Codes** [comment]
```python
from nicegui import ui, app

def on_disconnect():
    print("Browser tab closed or native mode closed. Shutting down server...")
    app.shutdown() 
    exit()

app.on_disconnect(on_disconnect)

ui.label("Hello NiceGUI! Close the tab to exit the script.")
ui.run(reload=True, native=True)
```

May this can help...

**Comment by evnchn** [comment]
To clarify things:

On 2.24.2 and latest main (~3.6.1), both versions correctly print `app shutdown`

The issue is that when click X, it doesn't bring down the app with it, which one kinda expects it to do.

**Comment by evnchn** [comment]
Yes clicking X to shut down the server only breaks under `reload=True` 

Yes this is Windows-specific (so @rodja's feeling is correct at https://github.com/zauberzeug/nicegui/issues/2107#issuecomment-1838199945). 

I tried it on macOS (no devcontainer for native mode dev) and it works. 

This is going to be a tough one. I took like 5 shots with Claude Opus 4.5. 

Get ready to just document this as a shortcoming of `reload=True` if persisted effort does not fix it.
